server/src-lib/Hasura/RQL/DDL/RemoteSchema/Permission.hs

```diff
@@ -0,0 +1,966 @@
+                True -> do
+                  variableName <- getVariableName
+                  pure $
+                    G.VVariable $
+                    SessionPresetVariable (mkSessionVariable t) gType variableName $
+                    SessionArgumentPresetEnum enumVals
+                False -> refute $ pure $ ExpectedEnumValue typeName value
+            _ -> refute $ pure $ ExpectedEnumValue typeName value
+        Just (PresetInputObject inputValueDefinitions) ->
+          let inpValsMap = mapFromL G._ivdName inputValueDefinitions
+              parseInputObjectField k val = do
+                inpVal <- onNothing (Map.lookup k inpValsMap) (refute $ pure $ KeyDoesNotExistInInputObject k typeName)
+                parsePresetValue (G._ivdType inpVal) k isStatic val
+          in
+          case value of
+            G.VObject obj ->
+              G.VObject <$> Map.traverseWithKey parseInputObjectField obj
+            _ -> refute $ pure $ ExpectedInputObject typeName value
+    G.TypeList _ gType' ->
+      case value of
+        G.VList lst -> G.VList <$> traverse (parsePresetValue gType' varName isStatic) lst
+        -- The below is valid because singleton GraphQL values can be "upgraded"
+        -- to array types. For ex: An `Int` value can be provided as input to
+        -- a type `[Int]` or `[[Int]]`
+        s'@(G.VString s) ->
+          case isSessionVariable s of
+            True -> refute $ pure $ DisallowSessionVarForListType varName
+            False -> parsePresetValue gType' varName isStatic s'
+        v -> parsePresetValue gType' varName isStatic v
+  where
+    -- We need to have distinct variable names to avoid clashes between
+    -- variable names.
+    getVariableName :: m G.Name
+    getVariableName = do
+      ctr <- get
+      modify (+ 1)
+      pure $ G.unsafeMkName (G.unName varName <> "__" <> T.pack (show ctr))
+
+parsePresetDirective
+  :: forall m
+  .  ( MonadValidate [RoleBasedSchemaValidationError] m
+     , MonadReader G.SchemaDocument m
+     , MonadState Int m
+     )
+  => G.GType
+  -> G.Name
+  -> G.Directive Void
+  -> m (G.Value RemoteSchemaVariable)
+parsePresetDirective gType parentArgName (G.Directive name args) = do
+  if | Map.null args -> refute $ pure $ NoPresetArgumentFound
+     | otherwise -> do
+         val <- onNothing (Map.lookup $$(G.litName "value") args) $ refute $ pure $ InvalidPresetArgument parentArgName
+         isStatic <-
+           case (Map.lookup $$(G.litName "static") args) of
+             Nothing -> pure False
+             (Just (G.VBoolean b)) -> pure b
+             _ -> refute $ pure $ InvalidStaticValue
+         parsePresetValue gType parentArgName isStatic val
+
+-- | validateDirective checks if the arguments of a given directive
+--   is a subset of the corresponding upstream directive arguments
+--   *NOTE*: This function assumes that the `providedDirective` and the
+--   `upstreamDirective` have the same name.
```

Comment:
I was hesitant to add this check initially because it will be a server error, the caller of this function should take care of this. But, I guess I can throw an `unexpected` error when the above situation happens!

Reply:
I have added a check for this